### PR TITLE
Maybe the last restore tweaks

### PR DESCRIPTION
**Pull request by czue**
possibly final (?) set of prerequisites for ownership cleanliness PR

major changes:
- pass project through to restores in a bunch more tests
- add back an `IndexTree` structure to manage case relevancy on phones (and tests)

wait for build

@snopoke @dannyroberts @emord

### CODE REVIEW

File: corehq/ex-submodules/casexml/apps/phone/tests/test_state_hash.py [hunk 1]
```diff
@@ -59,16 +62,18 @@ def testMismatch(self):
...
+        generate_restore_payload(
+            self.project, self.user, self.sync_log.get_id,
+            version=V2, state_hash=str(real_hash)
+        )
         
         try:
-            generate_restore_payload(self.user, self.sync_log.get_id,
-                                                version=V2, state_hash=str(bad_hash))
+            generate_restore_payload(self.project, self.user, self.sync_log.get_id,
+                                                  version=V2, state_hash=str(bad_hash))
```
Comment (dimagimon) [inline corehq/ex-submodules/casexml/apps/phone/tests/test_state_hash.py]: E127 continuation line over-indented for visual indent

**Comment by czue** [comment]
:white_check_mark: :smiling_imp:

**Comment by czue** [comment]
(this is safe to merge, though I just found an issue with the new logic that I'm looking into - nothing calls it yet though)

File: corehq/ex-submodules/casexml/apps/phone/dbaccessors/sync_logs_by_user.py [hunk 2]
```diff
@@ -11,3 +12,11 @@ def get_last_synclog_for_user(user_id):
         reduce=False,
         include_docs=True,
     ).one()
+
+
+def get_all_sync_logs_docs():
```
Comment (snopoke) [inline corehq/ex-submodules/casexml/apps/phone/dbaccessors/sync_logs_by_user.py]: Presumably this should only be used in tests. Might be worth adding an assert that `settings.UNIT_TESTING` is true.

File: corehq/ex-submodules/casexml/apps/phone/tests/test_index_tree.py [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+        # just pruning one child should preserve the parent index
+        tree.prune_case(child_id_1)
+        self.assertTrue(parent_id in tree.live_indices)
+        self.assertFalse(parent_id in tree.dependent_indices)
+        self.assertFalse(child_id_1 in tree.live_indices)
+        self.assertFalse(child_id_1 in tree.dependent_indices)
+        self.assertTrue(child_id_2 in tree.live_indices)
+        self.assertFalse(child_id_2 in tree.dependent_indices)
+
+        # pruning the other one should wipe it
```
Comment (snopoke) [inline corehq/ex-submodules/casexml/apps/phone/tests/test_index_tree.py]: Shouldn't `parent` still be in `live_indices`?

Comment (czue) [inline corehq/ex-submodules/casexml/apps/phone/tests/test_index_tree.py]: possibly - actually stop reviewing this for now - i think i need to rework it (again) :(

File: corehq/ex-submodules/casexml/apps/phone/models.py [hunk 4]
```diff
@@ -408,9 +496,57 @@ def phone_is_holding_case(self, case_id):
...
+        dependencies_not_to_remove = dependencies - self.dependent_case_ids_on_phone
+
+        def _remove_case(to_remove):
+            # uses closures for assertions
+            assert to_remove in self.dependent_case_ids_on_phone
+            indices = self.index_tree.indices.pop(to_remove, [])
+            if to_remove != case_id:
+                # if the case had indexes they better also be in our removal list (except for ourselves)
+                for index in indices:
+                    assert index in candidates_to_remove, "expected {} in {} but wasn't".format(index, candidates_to_remove)
```
Comment (dimagimon) [inline corehq/ex-submodules/casexml/apps/phone/models.py]: E501 line too long (124 > 115 characters)

File: corehq/ex-submodules/casexml/apps/phone/models.py [hunk 5]
```diff
@@ -408,9 +496,57 @@ def phone_is_holding_case(self, case_id):
...
+        if not dependencies_not_to_remove:
+            # this case's entire relevancy chain is in dependent cases
+            # this means they can all now be removed.
+            this_case_indices = self.index_tree.indices.get(case_id, [])
+            for to_remove in candidates_to_remove:
+                _remove_case(to_remove)
+
+            for this_case_index in this_case_indices:
+                if (this_case_index in self.dependent_case_ids_on_phone and
+                            this_case_index not in candidates_to_remove):
```
Comment (dimagimon) [inline corehq/ex-submodules/casexml/apps/phone/models.py]: E127 continuation line over-indented for visual indent

File: corehq/ex-submodules/casexml/apps/phone/tests/test_index_tree.py [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+        # just pruning one child should preserve the parent index
+        sync_log.prune_case(child_id_1)
+        self.assertTrue(grandparent_id in sync_log.case_ids_on_phone)
+        self.assertTrue(grandparent_id in sync_log.dependent_case_ids_on_phone)
+        self.assertTrue(parent_id in sync_log.case_ids_on_phone)
+        self.assertTrue(parent_id in sync_log.dependent_case_ids_on_phone)
+        self.assertFalse(child_id_1 in sync_log.case_ids_on_phone)
+
+
+        # pruning the other one should wipe it
```
Comment (dimagimon) [inline corehq/ex-submodules/casexml/apps/phone/tests/test_index_tree.py]: E303 too many blank lines (2)

File: corehq/ex-submodules/casexml/apps/phone/tests/test_index_tree.py [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+        sync_log.prune_case(peer_id_3)
+        for id in all_ids:
+            self.assertFalse(id in sync_log.case_ids_on_phone)
+
+    def test_prune_self_indexing(self):
+        [id] = ['recursive']
+        tree = IndexTree(indices={
+            id: [id],
+        })
+        sync_log =  SimplifiedSyncLog(index_tree=tree, case_ids_on_phone=set([id]))
```
Comment (dimagimon) [inline corehq/ex-submodules/casexml/apps/phone/tests/test_index_tree.py]: E222 multiple spaces after operator

File: corehq/ex-submodules/casexml/apps/phone/dbaccessors/sync_logs_by_user.py (hunk 2 above)
Comment (czue) [inline corehq/ex-submodules/casexml/apps/phone/dbaccessors/sync_logs_by_user.py]: good point, updated

**Comment by czue** [comment]
@snopoke I think this is ready for review again, though feel free to wait for build

**Comment by czue** [comment]
@snopoke all the major changes are now :white_check_mark: so probably ok to continue reviewing now

File: corehq/ex-submodules/casexml/apps/phone/tests/test_sync_mode.py [hunk 8]
```diff
@@ -1021,8 +1092,50 @@ def testComplicatedGatesBug(self):
...
+
+        # have another user modify the owner ID of the dependent case to be the shared ID
+        self.factory.create_or_update_cases([
+            CaseStructure(
+                case_id=parent_id,
+                attrs={'owner_id': SHARED_ID},
+            )],
+            form_extras={
+                'last_sync_token': None,
+            }
```
Comment (dimagimon) [inline corehq/ex-submodules/casexml/apps/phone/tests/test_sync_mode.py]: E123 closing bracket does not match indentation of opening bracket's line

File: corehq/ex-submodules/casexml/apps/phone/tests/test_index_tree.py [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+
+        # this should prune the child but not the parent
+        sync_log.prune_case(child_id)
+        self.assertFalse(child_id in sync_log.case_ids_on_phone)
+        self.assertTrue(parent_id in sync_log.case_ids_on_phone)
+        self.assertFalse(child_id in sync_log.dependent_case_ids_on_phone)
+        self.assertFalse(parent_id in sync_log.dependent_case_ids_on_phone)
+
+        # then pruning the parent should prune it
+        # this should prune the child but not the parent
```
Comment (snopoke) [inline corehq/ex-submodules/casexml/apps/phone/tests/test_index_tree.py]: rogue comment

File: corehq/ex-submodules/casexml/apps/phone/tests/test_sync_mode.py [hunk 10]
```diff
@@ -1021,8 +1092,50 @@ def testComplicatedGatesBug(self):
...
+            )
+
+    def test_dependent_case_becomes_relevant_at_sync_time(self):
+        """
+        Make a case that is only relevant through a dependency.
+        Then update it to be actually relevant.
+        Make sure the sync removes it from the dependent list.
+        """
+        # create a parent and child case (with index) from one user
+        parent_id, child_id = [uuid.uuid4().hex for i in range(2)]
```
Comment (snopoke) [inline corehq/ex-submodules/casexml/apps/phone/tests/test_sync_mode.py]: this line gets overwritten by the next line

**Comment by czue** [comment]
woohoo! thanks for thorough review.

addressed comments in https://github.com/dimagi/commcare-hq/pull/6759

**Pull request merged**